### DISCUSSION TITLE
Question Captive Portal

### DISCUSSION THREAD

**Discussion by crbyxwpzfl** [mathieucarbou/MycilaWebSerial#13] [Q&A] [answered]
Hi I really like your WebSerial and it works great! Thank you for this!

**tldr;**  How do I respond to any request with the WebSerial page? 

## more info
I am trying to use WebSerial as a captive portal. So you just connect to the access Point and the WebSerial Page pops up. 
To trigger the Captive Portal you have to have two things, I think.

#### a DNS-Server to redirects any domain name to you access Point ip 
done via `dnsServer.start(53, "*", WiFi.softAPIP());` and `void loop() { dnsServer.processNextRequest(); }` see [here](https://github.com/me-no-dev/ESPAsyncWebServer/blob/7f3753454b1f176c4b6d6bcd1587a135d95ca63c/examples/CaptivePortal/CaptivePortal.ino#L39) 

#### plus respond with the WebSerial page to any request. 
Initaly I thought `server.onNotFound([](AsyncWebServerRequest* request) { request->redirect("/webserial");  });` from [your example](https://github.com/mathieucarbou/MycilaWebSerial/blob/a1b2389a4da52892a8f3887dcdca74169be938e7/examples/Logging/Logging.ino#L40) is sufficient. but this does not trigger the Captive Portal. 

A workaround is to redirect to the WebSerial Page via html with `<meta http-equiv='refresh' content='0; url=http://ip/webserial' />` and pass this instead.
```c
  server.onNotFound([](AsyncWebServerRequest* request) { 
    AsyncResponseStream *response = request->beginResponseStream("text/html");
    response->printf("<!DOCTYPE html><html><meta http-equiv='refresh' content='0; url=http://%s/webserial' /><head><title>Captive Portal</title></head><body>", WiFi.softAPIP().toString().c_str());
    response->printf("<p>This is out captive portal front page. http://%s/webserial </p>", WiFi.softAPIP().toString().c_str());
    response->print("</body></html>");
    request->send(response);
    //request->redirect("/webserial");   // this does not work for the captive portal
  });
```

<br>

this html redirect triggers the Captive Portal but it seems overly complicated to me. Im not too good in programming and hope you have some tips for me how to directly provide the WebSerial Page.
Please close this at any point I understand that this is not really part of your library.

<br>

### full example
<details>
<summary>Logging.ino with Captive Portal Workaround tested on esp32s3 with ios</summary>

```c
#include <Arduino.h>
#include <AsyncTCP.h>
#include <WiFi.h>

#include <DNSServer.h>
#include <ESPAsyncWebServer.h>
#include <WString.h>
#include <MycilaWebSerial.h>

AsyncWebServer server(80);

DNSServer dnsServer;

static uint32_t last = millis();
static uint32_t count = 0;

void setup() {
  Serial.begin(115200);

  WiFi.softAP("WSLDemo");

  // by default DNSServer is started serving any "*" domain name. It will reply
  // AccessPoint's IP to all DNS request (this is required for Captive Portal detection)
  dnsServer.start(53, "*", WiFi.softAPIP());

  Serial.print("IP Address: ");
  Serial.println(WiFi.softAPIP().toString());

  WebSerial.onMessage([](const String& msg) { Serial.println(msg); });
  WebSerial.begin(&server);
  WebSerial.setBuffer(100);

  server.onNotFound([](AsyncWebServerRequest* request) { 
    AsyncResponseStream *response = request->beginResponseStream("text/html");
    response->printf("<!DOCTYPE html><html><meta http-equiv='refresh' content='0; url=http://%s/webserial' /><head><title>Captive Portal</title></head><body>", WiFi.softAPIP().toString().c_str());
    response->printf("<p>This is out captive portal front page. http://%s/webserial </p>", WiFi.softAPIP().toString().c_str());
    response->print("</body></html>");
    request->send(response);
    //request->redirect("/webserial");   // this does not work for the captive portal
  });

  server.begin();
}

void loop() {

  dnsServer.processNextRequest();  // to process dns requests

  if (millis() - last > 1000) {
    count++;

    WebSerial.print(F("IP address: "));
    WebSerial.println(WiFi.softAPIP());
    WebSerial.printf("Uptime: %lums\n", millis());
    WebSerial.printf("Free heap: %" PRIu32 "\n", ESP.getFreeHeap());

    last = millis();
  }
}
```

</details>

**Comment by mathieucarbou** [comment, marked as answer]
It is not meant to be used as a Captive portal. Have a look at [MycilaESPConnect/](https://oss.carbou.me/MycilaESPConnect/) for that.

**Reply by crbyxwpzfl** [reply]
thanks for the tip and quick reply I will use this.
